README.md
# digitalbits-hd-wallet

> 🔐 Key derivation for the DigitalBits blockchain (based on Stellar's [SEP-0005](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0005.md)) 🚀

<p align="center">
  <a href="https://www.npmjs.com/package/@bajetech/digitalbits-hd-wallet">
    <img alt="npm (scoped)" src="https://img.shields.io/npm/v/@bajetech/digitalbits-hd-wallet?style=for-the-badge">
  </a>
  <a href="https://nodejs.org">
    <img alt="Node.js" src="https://img.shields.io/badge/node->=12-yellowgreen?style=for-the-badge&labelColor=000000">
  </a>
  <a href="https://github.com/bajetech/digitalbits-hd-wallet/actions/workflows/pipeline.yml">
    <img alt="GitHub Workflow Status" src="https://img.shields.io/github/workflow/status/bajetech/digitalbits-hd-wallet/digitalbits-hd-wallet%20CI?label=GitHub%20Actions&logo=github&style=for-the-badge">
  </a>
</p>

## IMPORTANT NOTICE

**Update to v1.1.0 or higher**: Versions of this package prior to v1.1.0 are broken! It is very likely that if you attempt to use a pre-1.1.0 release of this package in your projects, they will NOT run! Please be sure to update to v1.1.0 or higher for a working release.

## Installation

```bash
yarn add @bajetech/digitalbits-hd-wallet  # or npm i @bajetech/digitalbits-hd-wallet
```

## Usage

```js
import DigitalBitsHDWallet from "@bajetech/digitalbits-hd-wallet";

const mnemonic = DigitalBitsHDWallet.generateMnemonic();
const wallet = DigitalBitsHDWallet.fromMnemonic(mnemonic);

wallet.getPublicKey(0); // => GDKYMXOAJ5MK4EVIHHNWRGAAOUZMNZYAETMHFCD6JCVBPZ77TUAZFPKT
wallet.getSecret(0); // => SCVVKNLBHOWBNJYHD3CNROOA2P3K35I5GNTYUHLLMUHMHWQYNEI7LVED
wallet.getKeypair(0); // => DigitalBitsBase.Keypair for account 0
wallet.derive(`m/44'/148'/0'`); // => raw key for account 0 as a Buffer

// wallet instance from seeds
const seedHex =
  "794fc27373add3ac7676358e868a787bcbf1edfac83edcecdb34d7f1068c645dbadba563f3f3a4287d273ac4f052d2fc650ba953e7af1a016d7b91f4d273378f";
const seedBuffer = Buffer.from(seedHex);
DigitalBitsHDWallet.fromSeed(seedHex);
DigitalBitsHDWallet.fromSeed(seedBuffer);

// mnemonics with different lengths
DigitalBitsHDWallet.generateMnemonic(); // 24 words
DigitalBitsHDWallet.generateMnemonic({ entropyBits: 224 }); // 21 words
DigitalBitsHDWallet.generateMnemonic({ entropyBits: 160 }); // 18 words
DigitalBitsHDWallet.generateMnemonic({ entropyBits: 128 }); // 12 words

// validate a mnemonic
DigitalBitsHDWallet.validateMnemonic("too short and non wordlist words"); // false
```

## Mnemonic Language

Mnemonics can be generated in any language supported by the underlying [bip39 npm module](https://github.com/bitcoinjs/bip39).

The full list of language keys are under exports 'wordlists' [here](https://github.com/bitcoinjs/bip39/blob/master/index.js).

### Usage

```js
import DigitalBitsHDWallet from "@bajetech/digitalbits-hd-wallet";

// traditional chinese - 24 words
DigitalBitsHDWallet.generateMnemonic({
  language: "chinese_traditional",
});
// => '省 从 唯 芽 激 顿 埋 愤 碳 它 炸 如 青 领 涨 骤 度 牲 朱 师 即 姓 讲 蒋'

// french - 12 words
DigitalBitsHDWallet.generateMnemonic({ language: "french", entropyBits: 128 });
// => 'directif terrible légume dérober science vision venimeux exulter abrasif vague mutuel innocent'
```

## Randomness

- NodeJs: crypto.randomBytes
- Browser: window.crypto.getRandomValues

(using [randombytes npm module](https://github.com/crypto-browserify/randombytes))

## TypeScript Support

Many thanks to [Kyle Roach](https://github.com/iRoachie) for converting this library to [TypeScript](https://www.typescriptlang.org) via the following [PR](https://github.com/bajetech/digitalbits-hd-wallet/pull/3). TypeScript support is now available as of version **1.2.0**.

## Tests

All [SEP-0005 test cases](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0005.md#test-cases) are exercised [here](https://github.com/bajetech/digitalbits-hd-wallet/blob/main/test/sep0005.test.js) against [these](https://github.com/bajetech/digitalbits-hd-wallet/tree/main/test/data).

## Credits

This package was cloned and adapted from [`stellar-hd-wallet`](https://github.com/chatch/stellar-hd-wallet) for use with the [DigitalBits blockchain network](https://digitalbits.io), which is itself adapted from the [Stellar blockchain network](https://www.stellar.org).

Specifically the following differences from `stellar-hd-wallet` are worthy of mention:

- GitHub Actions is used for a CI workflow instead of Travis CI.
- All _direct_ package dependencies and dev dependencies are brought up-to-date and the code modified where necessary to use up-to-date APIs.
- The `xdb-digitalbits-base` package is used instead of `stellar-base`.
- As of version 1.2.0 the package is written in TypeScript, with built-in TypeScript definitions exported.
- Yarn is used as the package manager of choice for development of this library.


CHANGELOG.md
# CHANGELOG

## v1.2.3

- build: bump dev dependencies [a2bdb4d](https://github.com/bajetech/digitalbits-hd-wallet/commit/a2bdb4d050de30db37a360bd1961d908d2c44c9c)

## v1.2.2

- build: bump dev dependencies [62e39ad](https://github.com/bajetech/digitalbits-hd-wallet/commit/62e39ade74f11b5f32c219372d3bab8b8cf884b8)

## v1.2.1

- refactor: use 'es6' for TS target and module config settings [2c9a483](https://github.com/bajetech/digitalbits-hd-wallet/commit/2c9a4831616348c281806912ba41267648669649)
- chore: switch from npm to Yarn [97d14da](https://github.com/bajetech/digitalbits-hd-wallet/commit/97d14da14943d3f9ec18c0e2b95e3103cfba70ad)

## v1.2.0

**NOTE**: This release marks the addition of TS support to this library.

The individual changes are as follows:

- feat: convert project to typescript. PR [#3](https://github.com/bajetech/digitalbits-hd-wallet/pull/3)
- chore: bump mocha dev dependency to 9.1.2 [e80d4a5](https://github.com/bajetech/digitalbits-hd-wallet/commit/e80d4a5f12b19951e253a035e75980d2e0c07f8f)

## v1.1.0

**NOTE**: This is a major bug-fix release (although only the minor release number is being given a bump). This is because there is no change to the API but the bug fixed is more serious than just releasing it as a patch update.

The individual changes are as follows:

- Adds @babel/plugin-transform-runtime as a devDependency, @babel/runtime as a dependency and updates babel configuration. [f6d3c1a](https://github.com/bajetech/digitalbits-hd-wallet/commit/f6d3c1acadc2d24724300ad3648f94f9308d9225)
- Changes source files extension from `.mjs` to `.js` [6553592](https://github.com/bajetech/digitalbits-hd-wallet/commit/655359237006b4a7f680adbeaf30324dd799cdc6)
- Updates tests so they will run given the change in src files from _.mjs to _.js [c8d3273](https://github.com/bajetech/digitalbits-hd-wallet/commit/c8d3273e74efc55b6b7f074557198540393634a9)

## v1.0.2

- NO code changes.
- README updated.

## v1.0.1

- Adds CHANGELOG.
- Fixes package-lock.json so it has the correct version number for the release.

## v1.0.0

This release comes with NO changes. Instead, we are simply bumping the version to 1.0.0 to indicate we are committed to following the [semantic versioning spec](http://semver.org) from this point on when publishing releases of this package.

## v0.0.10

Initial (official) release! 🎉

🔐 Key derivation for the DigitalBits blockchain (based on Stellar's [SEP-0005](https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0005.md)) 🚀

Future versions will keep package dependencies up-to-date and refine the library's API as needed to support any changes to SEP-0005 or the underlying [bip39 npm module](https://github.com/bitcoinjs/bip39).


test/sep0005.test.js
import assert from "assert";
import has from "lodash/has";
import DigitalBitsHDWallet from "../lib/digitalbits-hd-wallet";

const assertKeypair = (actualKeypair, expectedPublicKey, expectedSecret) => {
  assert.strictEqual(actualKeypair.publicKey(), expectedPublicKey);
  assert.strictEqual(actualKeypair.secret(), expectedSecret);
};

const specTestCase = (num) => () => {
  let testCase;
  let wallet;

  before(async function () {
    testCase = require(`./data/sep0005-testcase-${num}.js`);

    wallet = has(testCase, "passphrase")
      ? DigitalBitsHDWallet.fromMnemonic(testCase.seedWords, testCase.passphrase)
      : DigitalBitsHDWallet.fromMnemonic(testCase.seedWords);
  });

  it("derives expected parent key", () => {
    assert.strictEqual(wallet.derive(`m/44'/148'`).toString("hex"), testCase.parentKey);
  });

  it("derives expected child keys", () => {
    testCase.keypairs.forEach(([publicKey, secret], index) =>
      assertKeypair(wallet.getKeypair(index), publicKey, secret)
    );
  });
};

describe("SEP-0005", () => {
  describe("Test Case 1", specTestCase(1));
  describe("Test Case 2", specTestCase(2));
  describe("Test Case 3", specTestCase(3));
  describe("Test Case 4", specTestCase(4));
  describe("Test Case 5", specTestCase(5));
});


test/sep0005.test.mjs
import assert from "assert";
import has from "lodash/has.js";
import DigitalBitsHDWallet from "../src/digitalbits-hd-wallet.mjs";

const assertKeypair = (actualKeypair, expectedPublicKey, expectedSecret) => {
  assert.strictEqual(actualKeypair.publicKey(), expectedPublicKey);
  assert.strictEqual(actualKeypair.secret(), expectedSecret);
};

const specTestCase = (num) => () => {
  let testCase;
  let wallet;

  before(async function () {
    testCase = await import(`./data/sep0005-testcase-${num}.js`);
    testCase = testCase.default;

    wallet = has(testCase, "passphrase")
      ? DigitalBitsHDWallet.fromMnemonic(testCase.seedWords, testCase.passphrase)
      : DigitalBitsHDWallet.fromMnemonic(testCase.seedWords);
  });

  it("derives expected parent key", () => {
    assert.strictEqual(wallet.derive(`m/44'/148'`).toString("hex"), testCase.parentKey);
  });

  it("derives expected child keys", () => {
    testCase.keypairs.forEach(([publicKey, secret], index) =>
      assertKeypair(wallet.getKeypair(index), publicKey, secret)
    );
  });
};

describe("SEP-0005", () => {
  describe("Test Case 1", specTestCase(1));
  describe("Test Case 2", specTestCase(2));
  describe("Test Case 3", specTestCase(3));
  describe("Test Case 4", specTestCase(4));
  describe("Test Case 5", specTestCase(5));
});


test/data/sep0005-testcase-1.js
module.exports = {
  seedWords: "illness spike retreat truth genius clock brain pass fit cave bargain toe",
  parentKey: "e0eec84fe165cd427cb7bc9b6cfdef0555aa1cb6f9043ff1fe986c3c8ddd22e3",
  keypairs: [
    [
      "GDRXE2BQUC3AZNPVFSCEZ76NJ3WWL25FYFK6RGZGIEKWE4SOOHSUJUJ6",
      "SBGWSG6BTNCKCOB3DIFBGCVMUPQFYPA2G4O34RMTB343OYPXU5DJDVMN",
    ],
    [
      "GBAW5XGWORWVFE2XTJYDTLDHXTY2Q2MO73HYCGB3XMFMQ562Q2W2GJQX",
      "SCEPFFWGAG5P2VX5DHIYK3XEMZYLTYWIPWYEKXFHSK25RVMIUNJ7CTIS",
    ],
    [
      "GAY5PRAHJ2HIYBYCLZXTHID6SPVELOOYH2LBPH3LD4RUMXUW3DOYTLXW",
      "SDAILLEZCSA67DUEP3XUPZJ7NYG7KGVRM46XA7K5QWWUIGADUZCZWTJP",
    ],
    [
      "GAOD5NRAEORFE34G5D4EOSKIJB6V4Z2FGPBCJNQI6MNICVITE6CSYIAE",
      "SBMWLNV75BPI2VB4G27RWOMABVRTSSF7352CCYGVELZDSHCXWCYFKXIX",
    ],
    [
      "GBCUXLFLSL2JE3NWLHAWXQZN6SQC6577YMAU3M3BEMWKYPFWXBSRCWV4",
      "SCPCY3CEHMOP2TADSV2ERNNZBNHBGP4V32VGOORIEV6QJLXD5NMCJUXI",
    ],
    [
      "GBRQY5JFN5UBG5PGOSUOL4M6D7VRMAYU6WW2ZWXBMCKB7GPT3YCBU2XZ",
      "SCK27SFHI3WUDOEMJREV7ZJQG34SCBR6YWCE6OLEXUS2VVYTSNGCRS6X",
    ],
    [
      "GBY27SJVFEWR3DUACNBSMJB6T4ZPR4C7ZXSTHT6GMZUDL23LAM5S2PQX",
      "SDJ4WDPOQAJYR3YIAJOJP3E6E4BMRB7VZ4QAEGCP7EYVDW6NQD3LRJMZ",
    ],
    [
      "GAY7T23Z34DWLSTEAUKVBPHHBUE4E3EMZBAQSLV6ZHS764U3TKUSNJOF",
      "SA3HXJUCE2N27TBIZ5JRBLEBF3TLPQEBINP47E6BTMIWW2RJ5UKR2B3L",
    ],
    [
      "GDJTCF62UUYSAFAVIXHPRBR4AUZV6NYJR75INVDXLLRZLZQ62S44443R",
      "SCD5OSHUUC75MSJG44BAT3HFZL2HZMMQ5M4GPDL7KA6HJHV3FLMUJAME",
    ],
    [
      "GBTVYYDIYWGUQUTKX6ZMLGSZGMTESJYJKJWAATGZGITA25ZB6T5REF44",
      "SCJGVMJ66WAUHQHNLMWDFGY2E72QKSI3XGSBYV6BANDFUFE7VY4XNXXR",
    ],
  ],
};


src/hd-key.js
/**
 * This key derivation code was copied and adapted from:
 *   https://github.com/alepop/ed25519-hd-key/blob/master/src/index.ts
 * in order to remove js-nacl dependency which caused browser errors with
 * content-security-policy - see #12 for details. This modified implementation
 * is a pure JS implementation.
 *
 * The original ed25519-hd-key module is licensed under "GPL-3".
 *
 * This code is an almost exact copy of the code from:
 *   https://github.com/chatch/stellar-hd-wallet/blob/master/src/hd-key.js
 *
 * The only change was to convert the file to an ES module and to use an
 * ES6 import statement to import the `create-hmac/browser.js` module
 * instead of a CommonJS require statement.
 */

import createHmac from "create-hmac/browser";

const ED25519_CURVE = "ed25519 seed";
const HARDENED_OFFSET = 0x80000000;

export const derivePath = (path, seed) => {
  if (!isValidPath(path)) {
    throw new Error("Invalid derivation path");
  }
  const { key, chainCode } = getMasterKeyFromSeed(seed);
  const segments = path
    .split("/")
    .slice(1)
    .map(replaceDerive)
    .map((el) => parseInt(el, 10));
  return segments.reduce((parentKeys, segment) => CKDPriv(parentKeys, segment + HARDENED_OFFSET), {
    key,
    chainCode,
  });
};

const getMasterKeyFromSeed = (seed) => {
  const hmac = createHmac("sha512", ED25519_CURVE);
  const I = hmac.update(Buffer.from(seed, "hex")).digest();
  const IL = I.slice(0, 32);
  const IR = I.slice(32);
  return {
    key: IL,
    chainCode: IR,
  };
};

const CKDPriv = ({ key, chainCode }, index) => {
  const indexBuffer = Buffer.allocUnsafe(4);
  indexBuffer.writeUInt32BE(index, 0);
  const data = Buffer.concat([Buffer.alloc(1, 0), key, indexBuffer]);
  const I = createHmac("sha512", chainCode).update(data).digest();
  const IL = I.slice(0, 32);
  const IR = I.slice(32);
  return {
    key: IL,
    chainCode: IR,
  };
};

const replaceDerive = (val) => val.replace("'", "");
const pathRegex = new RegExp("^m(\\/[0-9]+')+$");
const isValidPath = (path) => {
  if (!pathRegex.test(path)) {
    return false;
  }
  return !path.split("/").slice(1).map(replaceDerive).some(isNaN);
};


src/hd-key.ts
/**
 * This key derivation code was copied and adapted from:
 *   https://github.com/alepop/ed25519-hd-key/blob/master/src/index.ts
 * in order to remove js-nacl dependency which caused browser errors with
 * content-security-policy - see #12 for details. This modified implementation
 * is a pure JS implementation.
 *
 * The original ed25519-hd-key module is licensed under "GPL-3".
 *
 * This code is an almost exact copy of the code from:
 *   https://github.com/chatch/stellar-hd-wallet/blob/master/src/hd-key.js
 *
 * The only change was to convert the file to an ES module and to use an
 * ES6 import statement to import the `create-hmac/browser.js` module
 * instead of a CommonJS require statement.
 */

import createHmac from "create-hmac/browser";

const ED25519_CURVE = "ed25519 seed";
const HARDENED_OFFSET = 0x80000000;

export const derivePath = (path: string, seed: string) => {
  if (!isValidPath(path)) {
    throw new Error("Invalid derivation path");
  }
  const { key, chainCode } = getMasterKeyFromSeed(seed);
  const segments = path
    .split("/")
    .slice(1)
    .map(replaceDerive)
    .map((el) => parseInt(el, 10));

  return segments.reduce((parentKeys, segment) => CKDPriv(parentKeys, segment + HARDENED_OFFSET), {
    key,
    chainCode,
  });
};

const getMasterKeyFromSeed = (seed: string) => {
  const hmac = createHmac("sha512", ED25519_CURVE);
  const I = hmac.update(Buffer.from(seed, "hex")).digest();
  const IL = I.slice(0, 32);
  const IR = I.slice(32);
  return {
    key: IL,
    chainCode: IR,
  };
};

const CKDPriv = ({ key, chainCode }: { key: Buffer; chainCode: Buffer }, index: number) => {
  const indexBuffer = Buffer.allocUnsafe(4);
  indexBuffer.writeUInt32BE(index, 0);
  const data = Buffer.concat([Buffer.alloc(1, 0), key, indexBuffer]);
  const I = createHmac("sha512", chainCode).update(data).digest();
  const IL = I.slice(0, 32);
  const IR = I.slice(32);
  return {
    key: IL,
    chainCode: IR,
  };
};

const replaceDerive = (val: string) => val.replace("'", "");
const pathRegex = new RegExp("^m(\\/[0-9]+')+$");

const isValidPath = (path: string) => {
  if (!pathRegex.test(path)) {
    return false;
  }

  return !path
    .split("/")
    .slice(1)
    .map(replaceDerive)
    .some((a) => isNaN(Number(a)));
};


src/hd-key.mjs
/**
 * This key derivation code was copied and adapted from:
 *   https://github.com/alepop/ed25519-hd-key/blob/master/src/index.ts
 * in order to remove js-nacl dependency which caused browser errors with
 * content-security-policy - see #12 for details. This modified implementation
 * is a pure JS implementation.
 *
 * The original ed25519-hd-key module is licensed under "GPL-3".
 *
 * This code is an almost exact copy of the code from:
 *   https://github.com/chatch/stellar-hd-wallet/blob/master/src/hd-key.js
 *
 * The only change was to convert the file to an ES module and to use an
 * ES6 import statement to import the `create-hmac/browser.js` module
 * instead of a CommonJS require statement.
 */

import createHmac from "create-hmac/browser.js";

const ED25519_CURVE = "ed25519 seed";
const HARDENED_OFFSET = 0x80000000;

export const derivePath = (path, seed) => {
  if (!isValidPath(path)) {
    throw new Error("Invalid derivation path");
  }
  const { key, chainCode } = getMasterKeyFromSeed(seed);
  const segments = path
    .split("/")
    .slice(1)
    .map(replaceDerive)
    .map((el) => parseInt(el, 10));
  return segments.reduce((parentKeys, segment) => CKDPriv(parentKeys, segment + HARDENED_OFFSET), {
    key,
    chainCode,
  });
};

const getMasterKeyFromSeed = (seed) => {
  const hmac = createHmac("sha512", ED25519_CURVE);
  const I = hmac.update(Buffer.from(seed, "hex")).digest();
  const IL = I.slice(0, 32);
  const IR = I.slice(32);
  return {
    key: IL,
    chainCode: IR,
  };
};

const CKDPriv = ({ key, chainCode }, index) => {
  const indexBuffer = Buffer.allocUnsafe(4);
  indexBuffer.writeUInt32BE(index, 0);
  const data = Buffer.concat([Buffer.alloc(1, 0), key, indexBuffer]);
  const I = createHmac("sha512", chainCode).update(data).digest();
  const IL = I.slice(0, 32);
  const IR = I.slice(32);
  return {
    key: IL,
    chainCode: IR,
  };
};

const replaceDerive = (val) => val.replace("'", "");
const pathRegex = new RegExp("^m(\\/[0-9]+')+$");
const isValidPath = (path) => {
  if (!pathRegex.test(path)) {
    return false;
  }
  return !path.split("/").slice(1).map(replaceDerive).some(isNaN);
};


src/digitalbits-hd-wallet.js
import has from "lodash/has";
import * as bip39 from "bip39";
import { derivePath } from "./hd-key";
import { Keypair } from "xdb-digitalbits-base";

const ENTROPY_BITS = 256; // = 24 word mnemonic

const INVALID_SEED = "Invalid seed (must be a Buffer or hex string)";
const INVALID_MNEMONIC = "Invalid mnemonic (see bip39)";

/**
 * Class for SEP-0005 key derivation.
 *
 * This code is copied and adapted from:
 *   https://github.com/chatch/stellar-hd-wallet/blob/b529d5ad19e9cc31029fd9fbae724856adf4b953/src/stellar-hd-wallet.js
 *
 * to use with the DigitalBits blockchain network and ecosystem.
 *
 * @see {@link https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0005.md|SEP-0005}
 */
class DigitalBitsHDWallet {
  /**
   * Instance from a BIP39 mnemonic string.
   * @param {string} mnemonic A BIP39 mnemonic
   * @param {string} [password] Optional mnemonic password
   * @param {string} [language='english'] Optional language of mnemonic
   * @throws {Error} Invalid Mnemonic
   */
  static fromMnemonic(mnemonic, password = undefined, language = "english") {
    if (!DigitalBitsHDWallet.validateMnemonic(mnemonic, language)) {
      throw new Error(INVALID_MNEMONIC);
    }
    return new DigitalBitsHDWallet(bip39.mnemonicToSeedSync(mnemonic, password).toString("hex"));
  }

  /**
   * Instance from a seed
   * @param {(string|Buffer)} binary seed
   * @throws {TypeError} Invalid seed
   */
  static fromSeed(seed) {
    let seedHex;

    if (Buffer.isBuffer(seed)) seedHex = seed.toString("hex");
    else if (typeof seed === "string") seedHex = seed;
    else throw new TypeError(INVALID_SEED);

    return new DigitalBitsHDWallet(seedHex);
  }

  /**
   * Generate a mnemonic using BIP39
   * @param {Object} props Properties defining how to generate the mnemonic
   * @param {Number} [props.entropyBits=256] Entropy bits
   * @param {string} [props.language='english'] name of a language wordlist as
   *          defined in the 'bip39' npm module. See module.exports.wordlists:
   *          here https://github.com/bitcoinjs/bip39/blob/master/index.js
   * @param {function} [props.rng] RNG function (default is crypto.randomBytes)
   * @throws {TypeError} Langauge not supported by bip39 module
   * @throws {TypeError} Invalid entropy
   */
  static generateMnemonic({
    entropyBits = ENTROPY_BITS,
    language = "english",
    rngFn = undefined,
  } = {}) {
    if (language && !has(bip39.wordlists, language))
      throw new TypeError(`Language ${language} does not have a wordlist in the bip39 module`);
    const wordlist = bip39.wordlists[language];
    return bip39.generateMnemonic(entropyBits, rngFn, wordlist);
  }

  /**
   * Validate a mnemonic using BIP39
   * @param {string} mnemonic A BIP39 mnemonic
   * @param {string} [language='english'] name of a language wordlist as
   *          defined in the 'bip39' npm module. See module.exports.wordlists:
   *          here https://github.com/bitcoinjs/bip39/blob/master/index.js
   * @throws {TypeError} Langauge not supported by bip39 module
   */
  static validateMnemonic(mnemonic, language = "english") {
    if (language && !has(bip39.wordlists, language))
      throw new TypeError(`Language ${language} does not have a wordlist in the bip39 module`);
    const wordlist = bip39.wordlists[language];
    return bip39.validateMnemonic(mnemonic, wordlist);
  }

  /**
   * New instance from seed hex string
   * @param {string} seedHex Hex string
   */
  constructor(seedHex) {
    this.seedHex = seedHex;
  }

  /**
   * Derive key given a full BIP44 path
   * @param {string} path BIP44 path string (eg. m/44'/148'/8')
   * @return {Buffer} Key binary as Buffer
   */
  derive(path) {
    const data = derivePath(path, this.seedHex);
    return data.key;
  }

  /**
   * Get DigitalBits account keypair for child key at given index
   * @param {Number} index Account index into path m/44'/148'/{index}
   * @return {xdb-digitalbits-base.Keypair} Keypair instance for the account
   */
  getKeypair(index) {
    const key = this.derive(`m/44'/148'/${index}'`);
    return Keypair.fromRawEd25519Seed(key);
  }

  /**
   * Get public key for account at index
   * @param {Number} index Account index into path m/44'/148'/{index}
   * @return {string} Public key
   */
  getPublicKey(index) {
    return this.getKeypair(index).publicKey();
  }

  /**
   * Get secret for account at index
   * @param {Number} index Account index into path m/44'/148'/{index}
   * @return {string} Secret
   */
  getSecret(index) {
    return this.getKeypair(index).secret();
  }
}

export default DigitalBitsHDWallet;


src/digitalbits-hd-wallet.mjs
import has from "lodash/has.js";
import bip39 from "bip39";
import { derivePath } from "./hd-key.mjs";
import { Keypair } from "xdb-digitalbits-base";

const ENTROPY_BITS = 256; // = 24 word mnemonic

const INVALID_SEED = "Invalid seed (must be a Buffer or hex string)";
const INVALID_MNEMONIC = "Invalid mnemonic (see bip39)";

/**
 * Class for SEP-0005 key derivation.
 *
 * This code is copied and adapted from:
 *   https://github.com/chatch/stellar-hd-wallet/blob/b529d5ad19e9cc31029fd9fbae724856adf4b953/src/stellar-hd-wallet.js
 *
 * to use with the DigitalBits blockchain network and ecosystem.
 *
 * @see {@link https://github.com/stellar/stellar-protocol/blob/master/ecosystem/sep-0005.md|SEP-0005}
 */
class DigitalBitsHDWallet {
  /**
   * Instance from a BIP39 mnemonic string.
   * @param {string} mnemonic A BIP39 mnemonic
   * @param {string} [password] Optional mnemonic password
   * @param {string} [language='english'] Optional language of mnemonic
   * @throws {Error} Invalid Mnemonic
   */
  static fromMnemonic(mnemonic, password = undefined, language = "english") {
    if (!DigitalBitsHDWallet.validateMnemonic(mnemonic, language)) {
      throw new Error(INVALID_MNEMONIC);
    }
    return new DigitalBitsHDWallet(bip39.mnemonicToSeedSync(mnemonic, password).toString("hex"));
  }

  /**
   * Instance from a seed
   * @param {(string|Buffer)} binary seed
   * @throws {TypeError} Invalid seed
   */
  static fromSeed(seed) {
    let seedHex;

    if (Buffer.isBuffer(seed)) seedHex = seed.toString("hex");
    else if (typeof seed === "string") seedHex = seed;
    else throw new TypeError(INVALID_SEED);

    return new DigitalBitsHDWallet(seedHex);
  }

  /**
   * Generate a mnemonic using BIP39
   * @param {Object} props Properties defining how to generate the mnemonic
   * @param {Number} [props.entropyBits=256] Entropy bits
   * @param {string} [props.language='english'] name of a language wordlist as
   *          defined in the 'bip39' npm module. See module.exports.wordlists:
   *          here https://github.com/bitcoinjs/bip39/blob/master/index.js
   * @param {function} [props.rng] RNG function (default is crypto.randomBytes)
   * @throws {TypeError} Langauge not supported by bip39 module
   * @throws {TypeError} Invalid entropy
   */
  static generateMnemonic({
    entropyBits = ENTROPY_BITS,
    language = "english",
    rngFn = undefined,
  } = {}) {
    if (language && !has(bip39.wordlists, language))
      throw new TypeError(`Language ${language} does not have a wordlist in the bip39 module`);
    const wordlist = bip39.wordlists[language];
    return bip39.generateMnemonic(entropyBits, rngFn, wordlist);
  }

  /**
   * Validate a mnemonic using BIP39
   * @param {string} mnemonic A BIP39 mnemonic
   * @param {string} [language='english'] name of a language wordlist as
   *          defined in the 'bip39' npm module. See module.exports.wordlists:
   *          here https://github.com/bitcoinjs/bip39/blob/master/index.js
   * @throws {TypeError} Langauge not supported by bip39 module
   */
  static validateMnemonic(mnemonic, language = "english") {
    if (language && !has(bip39.wordlists, language))
      throw new TypeError(`Language ${language} does not have a wordlist in the bip39 module`);
    const wordlist = bip39.wordlists[language];
    return bip39.validateMnemonic(mnemonic, wordlist);
  }

  /**
   * New instance from seed hex string
   * @param {string} seedHex Hex string
   */
  constructor(seedHex) {
    this.seedHex = seedHex;
  }

  /**
   * Derive key given a full BIP44 path
   * @param {string} path BIP44 path string (eg. m/44'/148'/8')
   * @return {Buffer} Key binary as Buffer
   */
  derive(path) {
    const data = derivePath(path, this.seedHex);
    return data.key;
  }

  /**
   * Get DigitalBits account keypair for child key at given index
   * @param {Number} index Account index into path m/44'/148'/{index}
   * @return {xdb-digitalbits-base.Keypair} Keypair instance for the account
   */
  getKeypair(index) {
    const key = this.derive(`m/44'/148'/${index}'`);
    return Keypair.fromRawEd25519Seed(key);
  }

  /**
   * Get public key for account at index
   * @param {Number} index Account index into path m/44'/148'/{index}
   * @return {string} Public key
   */
  getPublicKey(index) {
    return this.getKeypair(index).publicKey();
  }

  /**
   * Get secret for account at index
   * @param {Number} index Account index into path m/44'/148'/{index}
   * @return {string} Secret
   */
  getSecret(index) {
    return this.getKeypair(index).secret();
  }
}

export default DigitalBitsHDWallet;
